src/main/java/com/minecolonies/core/client/gui/WindowAssignCitizen.java

```diff
@@ -164,33 +191,40 @@ private void setupSettings(final Button settingsButton)
      */
     private void updateCitizens()
     {
-        citizens.clear();
-        citizens.addAll(colony.getCitizens().values());
-
-        //Removes all citizens which already have a job.
-        citizens = colony.getCitizens().values().stream()
-                     .filter(cit -> !(colony.getBuilding(cit.getHomeBuilding()) instanceof AbstractBuildingGuards.View))
-                     .sorted(Comparator.comparing((ICitizenDataView cit) -> (cit.getHomeBuilding() != null && cit.getHomeBuilding().equals(building.getPosition())) ? 0 : 1)
-                       .thenComparing(cit -> cit.getHomeBuilding() == null ? 0 : 1)
+        //Removes citizens that work from home and remove citizens already living here.
+        unassignedCitizens = colony.getCitizens().values().stream()
+                     .filter(cit -> !Objects.equals(cit.getHomeBuilding(), cit.getWorkBuilding()) && !building.getPosition().equals(cit.getHomeBuilding()))
+                     .sorted(Comparator.comparing((ICitizenDataView cit) -> cit.getHomeBuilding() == null ? 0 : 1)
                                .thenComparingLong(cit -> {
                                    if (cit.getWorkBuilding() == null)
                                    {
-                                       return 0;
+                                       if (cit.getHomeBuilding() == null)
+                                       {
+                                           return 0;
+                                       }
+                                       return Integer.MAX_VALUE;
                                    }
 
                                    return BlockPosUtil.getDistance2D(cit.getWorkBuilding(), building.getPosition());
```

Comment:
change this to normal 3d distance btw, our dist2d does manhatten dist which can sometimes be off